core/imt/inc/ROOT/RTaskArena.hxx

```diff
@@ -0,0 +1,79 @@
+// @(#)root/thread:$Id$
+// // Author: Xavier Valls Pla   08/05/20
+//
+/*************************************************************************
+ * Copyright (C) 1995-2020, Rene Brun and Fons Rademakers.               *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+//////////////////////////////////////////////////////////////////////////
+//                                                                      //
+// RTaskArena                                                           //
+//                                                                      //
+// This file implements the method to initialize and retrieve ROOT's    //
+// global task arena, together with a method to check for active        //
+// CPU bandwith control, and a class to wrap the tbb task arena with    //
+// the purpose of keeping tbb off the installed headers                 //
+//                                                                      //
+//////////////////////////////////////////////////////////////////////////
+
+#ifndef ROOT_RTaskArena
+#define ROOT_RTaskArena
+
+#include "RConfigure.h"
+#include <memory>
+
+// exclude in case ROOT does not have IMT support
+#ifndef R__USE_IMT
+// No need to error out for dictionaries.
+# if !defined(__ROOTCLING__) && !defined(G__DICTIONARY)
+#  error "Cannot use ROOT::Internal::RTaskArenaWrapper without defining R__USE_IMT."
+# endif
+#else
+
+/// tbb::task_arena is an alias of tbb::interface7::task_arena, which doesn't allow
+/// to forward declare tbb::task_arena without forward declaring tbb::interface7
+namespace tbb{
+namespace interface7{class task_arena;}
+using task_arena = interface7::task_arena;
+}
+
+namespace ROOT {
+namespace Internal {
+
+////////////////////////////////////////////////////////////////////////////////
+/// Wrapper for tbb::task_arena.
+///
+/// Necessary in order to keep tbb away from ROOT headers.
+/// This class is thought out to be used as a singleton.
+////////////////////////////////////////////////////////////////////////////////
+class RTaskArenaWrapper {
+public:
+   RTaskArenaWrapper(unsigned maxConcurrency = 0);
```

Comment:
Maybe I'm doing something not so nice, but this allows me to call RTaskArena::TaskArenaSize and return fNWorkers when there is no instance of RTaskArena. This was the previous behaviour of TPoolManager and it is useful f.e. when calling GetThreadPoolSize if the task arena is not active.

Reply:
your comment refers to another thread (i'll call it thread 2.) that for some reason github does not make "commentable".

original thread 2. comment:

> static implies you can have multiple RTaskArenaWrapper objects but they must have the same fNWorkers. But by design you can't have multiple RTaskArenaWrappers, so I would argue that static is actually misleading.
>
> can fNWorkers and TaskArenaSize be normal class members?

now about your reply to 2.: so what you actually want is a way to check whether the RTaskArenaWrapper exists or not, which `GetGlobalTaskArena` does not let you do. uhm ok, i don't really have a proposal that is not basically the same thing but different. I will suggest a comment for the static getter.

Reply:
@eguiraud  Even if I friend the factory function what I actually need is to friend make_shared, the way it's programmed right now. Found some workarounds for this (https://stackoverflow.com/a/25069711) but they still leave a public way to call the constructor (in this case through a shared_ptr returning call to the constructor)

That said I could always initialize the shared_ptr without std::make_shared, i.e. `std::shared_ptr<RTaskArena> sp(new RTaskArena(n)` which would introduce one more allocation, but this will only happen "from grapes to pears." Looks like a good solution, am I missing something?

Reply:
`std::shared_ptr<RTaskArena> sp(new RTaskArena(n)` is perfectly fine imo